design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+           data ksqlDB can work with. Extending the set of key formats ksqlDB supports immediately
+           opens up the use of ksqlDB with previously incompatible datasets.
+           
+## Motivation and background
+
+Data stored in Kafka has a key and a value. These can be serialized using different formats, but are
+generally use a common format. ksqlDB supports multiple _value_ [data formats][1], but requires the 
+key data format to be the `KAFKA`.  
+
+This limitation is particularly problematic for tables. ksqlDB is unable to access changelog topics
+in Kafka that have non-`KAFKA` formatted keys. As the key of the Kafka record is the `PRIMARY KEY` 
+of the table, it is essential that the key can be read if the changelog is to be materialised into a 
+table. When changelog topics have non-`KAFKA` key formats, the limitation precludes ksqlDB as a solution.
+
+The limitation is less damaging for streams. However, it is still the case that the user looses the
+ability to access the data in the Kafka record's key. If this data is not duplicated in the record's
+value, which generally seems to be the case, then the data is not accessible at all. If the data is
+required, then the limitation precludes ksqlDB as a solution.
+
+As well as unsupported _input_ key formats, ksqlDB is equally precluded should a solution require the 
+_output_ to have a non-`KAFKA` key format. ksqlDB is already often used as the glue between 
+disparate systems, even though it is limited to changing the _values_ format and structure.
+Supporting other key formats opens this up to also transforming the key into a different format. 
+
+In some cases users are able to work around this limitation. This may involve changing upstream code,
+or introducing pre-processing, or, in the case of Connect, using SMTs to convert the key format. All
+such solutions tend to increase the complexity of the system, and generally hurt performance.
+
+To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 
+
+## What is in scope
+
+ * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+ * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+ * Addition of new server configuration to provide defaults for key and value formats.
+ * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+   the server configuration provides a default.
+ * Support for additional key column data types, where the key format supports it:
+    * `DECIMAL`
+    * `BOOLEAN`
+ * Support of the following key formats:
+    * `DELIMITED`: single key columns as a single string value.
+    * `JSON` / `JSON_SR`: single key column as an anonymous value, i.e. not within a JSON object.
+    * `AVRO`: single key column as an anonymous value, i.e. not within an Avro record.
```

Comment:
If we have structured keys in the future, how can a user set a non-primitive single-column AVRO key format? -- Would it be better to reserve `AVRO` for this case? and use `PRIMITIVE_AVRO` (or similar) for the case you describe? (Similar for JSON actually).

Reply:
We already support `WRAP_SINGLE_VALUE` in the `WITH` clause to handle switching between anonymous and wrapped serialized forms.   The future work to support wrapped keys will introduce `WRAP_SINGLE_KEY`. 

I've added details to the KLIP.

Reply:
See https://github.com/confluentinc/ksql/pull/6017/files#diff-f01d9cdc48d983b2fc1e805fbac19028R287-R324

Reply:
Thanks. Seems I missed the memo. `WRAP_SINGLE_VALUE` sounds a little awkward to me personally...